### DISCUSSION TITLE
What `cabal install` did when type this command?

### DISCUSSION THREAD

**Discussion by WendaoLee** [haskell/cabal#9210] [Q&A] [answered]
🙇‍♂️I'm new to Haskell and cabal.And these are my questions when using with cabal.I do hope these noob questions won't burden you and  I would be gratefully appreciate for your answers.

My question is: **What `cabal install` did when type this command?**

For the package manager,I'm familiar with `npm` of `nodejs`.It seems like `cabal install remote-package` won't modify the `.cabal` like the `npm install remote-package`  will modify the `package.json`.And it not download files into the workspace like npm.

I notice `cabal` manage its package like that:

 1. When enter `cabal install remote-package`，it will download sourcecode from hackage into local directory `./cabal/packages` if not exist.
 2. Then try to complie it into `.hi` files for better performence,which will help for building project afterwards.
 3. The last thing is what I get confused:Finished above things,it will try to compile the download-package into executable file no matter it is just a lib?For instance,I try to install a lib named `sqilte-simple` with `cabal install sqlite-simple`.When all things done,it throw error:
 ```
Error: cabal-3.10.1.0.exe: Cannot build the executables in the package
sqlite-simple because it does not contain any executables. Check the .cabal
file for the package and make sure that it properly declares the components
that you expect.
```

Is that a useless design?Or has some reasons for it?🙇‍♂️I'm really curious about it.

And for just enter `cabal install` in a local cabal project.It seem that it will try to compile the project into executeable file and install it on `./cabal/bin`.

Is there any wrong about my cognition?🙇‍♂️And it would be delightful for someone could answer me the details of how this command actually work.

**Comment by ffaf1** [comment]
> And for just enter cabal install in a local cabal project.It seem that it will try to compile the project into executeable file and install it on `./cabal/bin`.

For now! Once you switch to a more modern `cabal`, it will be based on XDG paths (e.g. `~/.local/bin`).

**Comment by ffaf1** [comment]
Reading this again, it seems to be that your congnition is spot on (but wait for more experienced people)!

I would notice that `cabal install` is indeed for executables, while `cabal install --lib` (provisional) is for libraries (plain is to split it: see #6481).

**Reply by WendaoLee** [reply]
Wow，🥰thank you for your detailed description.

Well,I have a new question when using `cabal`.That is,some package (e.g a package named `http-client`) I used has  its build-dependences(e.g has dependence `http-types`).And it seems like that cabal won't automatically import these dependences unless I specified it in my `.cabal` file?(e.g I want to use a module in `http-types` but get modules not found error,though I specified `http-client` in `.cabal` file which has build-dependences of `http-types`).

All in all,sincerely appreciate for your help in today and yesterday!😽

**Reply by ffaf1** [reply]
> That is,some package (e.g a package named http-client) I used has its build-dependences(e.g has dependence http-types).And it seems like that cabal won't automatically import these dependences unless I specified it in my .cabal file?

That is correct. E.g.:
1. package `foo` (which depends on `bar`) exports constructor `Foo`;
2. package `bar` exports constructor `Bar`;
3. if you need to use *both* `Foo` and `Bar` in your package, you cannot merely state `build-depends: foo`, but need a more principled `build depends: foo, bar`

This might feel awkward at first, but imagine what a headache for you it would be tomorrow `foo` stopped depending on `bar` and you had to hunt down missing dependencies.

**Reply by WendaoLee** [reply]
Thanks a lot！🥰🥰🥰

**Comment by fgaz** [comment, marked as answer]
> For the package manager,I'm familiar with `npm` of `nodejs`.It seems like `cabal install remote-package` won't modify the `.cabal` like the `npm install remote-package` will modify the `package.json`.And it not download files into the workspace like npm.

Yes, the `install` command of `cabal` is very different from the `npm` one. To do the equivalent of `npm install` you don't have to run `cabal install` at all. Instead, you add the package you want to use to the `build-depends` section of the `.cabal` file of your package. Cabal will take care of fetching and building the dependency as needed when you run `cabal build`, `cabal run`, or other similar commands.

`cabal install` instead is more like `npm install --global`. It's mostly used by end users to install executables into their `~/.local/bin` (or equivalent) directory, or by advanced developers to create custom environments through the `--lib` flag.

**Reply by WendaoLee** [reply]
Thanks😘
